source/server/listener_manager_impl.cc

```diff
@@ -309,6 +310,23 @@ ListenerManagerStats ListenerManagerImpl::generateStats(Stats::Scope& scope) {
 bool ListenerManagerImpl::addOrUpdateListener(
     const envoy::config::listener::v3alpha::Listener& config, const std::string& version_info,
     bool added_via_api) {
+
+  // TODO(junr03): currently only one ApiListener can be installed via bootstrap to avoid having to
+  // build a collection of listeners, and to have to be able to warm and drain the listeners. In the
+  // future allow multiple ApiListeners, and allow them to be created via LDS as well as bootstrap.
```

Comment:
Exactly. That was the first approach that I tackled this with. I spent a few hours toying with different implementations. However, there was a good amount of refactoring required to get us there. Mostly around the API surface that the listener manager currently uses to interact with the listener, and how most of it was not applicable to API listeners. This meant that I needed to start messing more than I wanted with the listener manager code.

So the conclusion that I arrived at was this simple approach to get something in place. I agree that we can then build up to the Listener manager owning an interface rather than the implementations.